File: credit.c
//~/pset1/credit
//credit.c
//pset1 cs50
//onelmv
//9.15.2018
//verification if a credit card number is valid according to Luhn’s algorithm.

#include<stdio.h>
#include<cs50.h>
#include<math.h>

int n_length(long long cardNumb);
int luhn(long long cardNumb);
int card_name(long long cardNumb, int length);

int main(void)
{
    long long card_n;

    //prompt
    do
    {
        card_n = get_long_long("Number: ");
    }
    while (card_n <= 0);

    //Check if it is a correct number or not. If it is ,we use it.
    int l = n_length(card_n); //store number length.
    if (l >= 13 && l <= 16)
    {
        if (l == 14)
        {
            printf("INVALID\n");
        }

        //Luhn’s algorithm.
        if (luhn(card_n) == 1)
        {
            //Card type.
            int c = card_name(card_n, l);
            if (c == 1)
            {
                printf("VISA\n");
            }
            else if (c == 2)
            {
                printf("MASTERCARD\n");
            }
            else if (c == 3)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n"); // Invalid checksum.
        }
    }
    else
    {
        printf("INVALID\n");  //Invalid number length.
    }

}//end

//length number function.
int n_length(long long cardNumb)
{
    int count = 0;
    //This while go to check each digit (From the end to the beginning)
    while (cardNumb >= 1)
    {
        cardNumb = cardNumb / 10 ;
        count++;
    }
    return count;
}

//Luhn function
int luhn(long long cardNumb)
{
    int sum = 0;
    int temp = 0;

    while (cardNumb > 0)
    {
        sum = sum + cardNumb % 10;  //Store the last digit.
        cardNumb = cardNumb / 10;   //go to another digit.
        temp = (cardNumb % 10) * 2; //Multiply by two and store temporarily

        if (temp >= 10) //If the result is greater than 10, sum each digit.
        {
            temp = (temp % 10) + (temp / 10);
        }
        sum = sum + temp;
        cardNumb = cardNumb / 10;
    }
    if (sum % 10 == 0)  //If checksum is correct return 1.
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

//CArd type function
int card_name(long long cardNumb, int length)
{
    if (length == 16)  //Check if is MASTERCARD or VISA.
    {
        int temp = cardNumb / pow(10, 15); //take the first digit .. pow(10,15) = 10^15
        int temp2 = cardNumb / pow(10, 14); //take two first digit
        if (temp == 4)
        {
            return 1;  //VISA
        }
        else if (temp2 == 51 || temp2 == 52 || temp2 == 53 || temp2 == 54 || temp2 == 55)
        {
            return 2; // MASTERCARD
        }
    }
    else if (length == 13)
    {
        int temp3 = cardNumb / pow(10, 12);
        if (temp3 == 4)
        {
            return 1; //visa
        }

    }
    else if (length == 15)
    {
        int temp4 = cardNumb / pow(10, 13);
        if (temp4 == 34 || temp4 == 37)
        {
            return 3; //AMEX
        }
    }

    return 0; //INVALID
}

File: credit.py
# ~/pset6/credit
# credit.py
# pset6 cs50
# onelmv
# 12.22.2018
# verification if a credit card number is valid according to Luhn’s algorithm.

from cs50 import get_int
import math

# luhn algorithm function


def luhn_algorithm(card):

    # take numbers in pair's indexs starting by second-to-last digit
    # (storing in reverse)
    sub_list = card[(len(card)-2)::-2]
    sub_list_2 = card[::-2]

    # multiplying by 2 each number
    for i in range(len(sub_list)):
        # store each number multiplied
        n = sub_list[i] * 2

        # if the new number have 2 digit , the two are added.
        if n >= 10:
            sub_list[i] = (n % 10) + (n // 10)
        else:
            sub_list[i] = n

    # adding up each number of the two sub lists
    tmp = sum(sub_list) + sum(sub_list_2)

    # cheking if tmp end in 0
    if(tmp % 10) == 0:
        return True
    else:
        return False

# function for print card name


def name(card, leng):

    if leng == 16:
        if card[0] == 4:
            print("VISA")
        elif card[0] == 5 and (card[1] >= 1 and card[1] <= 5):  # if the 2 first digit are betewn 51 and 55
            print("MASTERCARD")
    elif leng == 13:
        if card[0] == 4:
            print("VISA")
    elif leng == 15:
        if card[0] == 3 and (card[1] == 4 or card[1] == 7):
            print("AMEX")

###################################################################


# promt
while True:
    card = get_int("Card number: ")
    if card >= 0:
        break

# converting the number(card) into a list
num_list = []
for digit in str(card):
    num_list.append(int(digit))

# lenght of th list
leng = len(num_list)

if leng >= 13 and leng <= 16:
    if leng == 14:
        print("INVALID")

    # Luhn’s algorithm.
    elif luhn_algorithm(num_list) == True:
        # printing card names
        name(num_list, leng)

    else:
        print("INVALID")
else:
    print("INVALID")